Keep track of MVCC properties in store

v01dstar opened the pull request:
<!--
Thank you for contributing to TiKV!

If you haven't already, please read TiKV's [CONTRIBUTING](https://github.com/tikv/tikv/blob/master/CONTRIBUTING.md) document.

If you're unsure about anything, just ask; somebody should be along to answer within a day or two.

PR Title Format:
1. module [, module2, module3]: what's changed
2. *: what's changed
-->

### What is changed and how it works?
<!--
Please create an issue first to describe the problem.

There MUST be one line starting with "Issue Number:  " and 
linking the relevant issues via the "close" or "ref".

For more info, check https://github.com/tikv/tikv/blob/master/CONTRIBUTING.md#linking-issues.
-->

Issue Number: Close #xxx

<!--
You could use "commit message" code block to add more description to the final commit message.
For more info, check https://github.com/tikv/tikv/blob/master/CONTRIBUTING.md#format-of-the-commit-message.
-->
What's Changed:

```commit-message

```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests <!-- At least one of them must be included. -->

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note
<!-- 
Compatibility change, improvement, bugfix, and new feature need a release note.

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

If you don't think this PR needs a release note then fill it with None.
If this PR will be picked to release branch, then a release note is probably required.
-->

```release-note

```

v01dstar committed:
Titan: Add a regression test for Titan disabling (#18477)

close tikv/tikv#18434

The real bug has been fixed by distinguishing Titan blob index from RocksDB blob index during RocksDB upgrade effort.
This PR just adds a regression test to prove it is fixed.
We reproduced the error with the same test code in 7.5

Signed-off-by: Yang Zhang <yang.zhang@pingcap.com>

3pointer committed:
backup_stream: encode ts related field into meta file path (#18482)

close tikv/tikv#18497, ref pingcap/tidb#61318

backup_stream: encode ts related field into meta file path.

Signed-off-by: 3pointer <luancheng@pingcap.com>
Signed-off-by: 3pointer <qdlc2010@gmail.com>

Co-authored-by: 山岚 <36239017+YuJuncen@users.noreply.github.com>

bufferflies committed:
pd: degrade useless log (#18504)

close tikv/tikv#18506

Signed-off-by: bufferflies <1045931706@qq.com>

YuJuncen committed:
backup: move verbose errors to warn level (#18494)

close tikv/tikv#18493

Some of non-fatal error level logs during backing up are now warn level.

Signed-off-by: Juncen Yu <yujuncen@pingcap.com>

overvenus committed:
sst_importer: allow write during ingesting sst (#18514)

ref tikv/tikv#18081

Signed-off-by: Neil Shen <overvenus@gmail.com>

v01dstar committed:
Titan: Fix turning off compatibility issue (#18495)

close tikv/tikv#18434

Fixing Titan blob indices causing snapshot apply failures after Titan is turned off bug.

Signed-off-by: Yang Zhang <yang.zhang@pingcap.com>

LykxSassinator committed:
region-size: Inherit relevant configs on region-size from the last config file. (#18505)

close tikv/tikv#18503

Make TiKV can inherit the last configurations on `region-size` to avoid change the default size
of region unexpectedly.

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

Leavrth committed:
sst_importer: spawn switch mode (#18542)

close tikv/tikv#18541

Run RPC function switch_mode on the blocking acceptable thread.

Signed-off-by: Jianjun Liao <jianjun.liao@outlook.com>

LykxSassinator committed:
raftstore: fix incorrect and misleading index logging in StoreMsg of slowlog. (#18562)

close tikv/tikv#18561

Fix incorrect and misleading index logging in StoreMsg of slowlog.

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

v01dstar committed:
Keep track of MVCC properties in store

Signed-off-by: Yang Zhang <yang.zhang@pingcap.com>

LykxSassinator committed:
raftstore: break up awaken regions into small batch. (#18544)

close tikv/tikv#18533

To mitigate the impact of stalls when awakening too many regions, we break up all regions into small batches.

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

LykxSassinator committed:
raftstore: move the handling of `CompactedEvent` to split-check worker. (#18565)

close tikv/tikv#18532

Optimize the handling of `CompactedEvent` in raftstore by moving it to `split-check` worker.

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

v01dstar committed:
Use sum instead of sum rate for MVCC on dashboard

Signed-off-by: Yang Zhang <yang.zhang@pingcap.com>

Dog-Du committed:
build: fix Error occurred while make doc (#18574)

close tikv/tikv#18573

fix Issue #18573 : Error occurred while make doc
fix it by modify Makefile
You can see the solution in the issue

Signed-off-by: DogDu <du.wen.hao@outlook.com>

Co-authored-by: DogDu <du.wen.hao@outlook.com>

LykxSassinator committed:
*: remove the logging for "sst ingest is too slow" to avoid latency jitters. (#18590)

close tikv/tikv#18549

Removes the logging for "sst ingest is too slow" to avoid latency jitters.

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

bufferflies committed:
log:degrade some unused error log to warning log (#18568)

close tikv/tikv#18506

replace some error log with warning log

Signed-off-by: bufferflies <1045931706@qq.com>

exit-code-1 committed:
metrics: Register Missing TiKV Configs to Prometheus (#18589)

close tikv/tikv#10047

Register Missing TiKV Configs to Prometheus
Due to recent iterations and refactoring in TiKV, some important module configurations were not reported via metrics. 
This PR registers the configuration metrics for these key modules

Signed-off-by: exit-code-1 <20194550@qq.com>
Signed-off-by: zhy <55792392+exit-code-1@users.noreply.github.com>

Co-authored-by: lucasliang <nkcs_lykx@hotmail.com>

ti-chi-bot committed:
build: bump tikv pkg version to 9.0.0-beta.2 (#18596)

ref tikv/tikv#15990

build: bump tikv pkg version

Signed-off-by: ti-chi-bot <ti-community-prow-bot@tidb.io>

LykxSassinator committed:
config: avoid inherting unnecessary configs to RaftstoreConfig from the last config file (#18626)

ref tikv/tikv#18503

Avoid the inheritage of unexpected configurations from the last config file.

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

LykxSassinator committed:
*: update raft-engine to optimize `fetch_entries_to`. (#18617)

close tikv/tikv#18605

Optimizes `fetch_entries_to` in Raft-Engine to reduce contention and improve performance under mixed workloads.

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

LykxSassinator commented on components/raftstore/src/store/msg.rs:
```diff
@@ -978,6 +977,12 @@ where
         abnormal_stores: Vec<u64>,
     },
 
+    DoneCollectWholeRangeMVCCStats {
```
How about
```suggestion
    UpdateWholeRangeMVCCStats {
```

LykxSassinator commented on components/raftstore/src/store/worker/compact.rs:
```diff
@@ -378,17 +391,31 @@ where
                 }
             }
+            Task::CollectWholeRangeMVCCStats { callback } => {
+                match self.engine.get_whole_range_stats(CF_WRITE) {
```
When `write_cf` contains numerous SST files, the compact worker may become overloaded, potentially delaying `mvcc_stats `updates during `get_whole_range_stats`.

Maybe we should consider concurrent access and updates to mitigate this bottleneck ?

v01dstar committed:
GC: remove compact on split (#18500)

close tikv/tikv#18451

Remove compaction scheduling on split key not found error. 

Split key not found error is usually returned when a region only has a few keys but a large number of stale MVCC versions.
This was the incentive of scheduling a compaction on split error. However, this solution brings a more serious problem, i.e. worker threads starvation.
We don't really need this extra compaction, we expect that, after the compaction optimization is done, this MVCC accumulation issue would be mitigated. So, once the "split key not found" occurs, we just rely on the CheckThenCompact to resolve it.

Signed-off-by: Yang Zhang <yang.zhang@pingcap.com>

hbisheng committed:
titan: Estimate raw blob size based on Titan metrics (#18628)

close tikv/tikv#18600

This PR introduces a mechanism to estimate Titan blob raw size based on
metrics collected from Titan during flush. 

A sliding window of recent (raw, compressed) size samples is used to 
compute an average expansion factor. This factor is then used to 
estimate the raw size of a blob when building SST table properties. 
This approach assumes that the compression ratio remains relatively 
stable for the workload.

Apparently this is not a perfect solution, but a workaround. A more 
direct solution—recording the uncompressed size in the blob index 
format—was previously proposed[1]. However, changing the data format 
would break backward compatibility and prevent rollbacks, which can be 
risky.

[1] https://github.com/tikv/titan/pull/334#issuecomment-3026813148

Signed-off-by: Bisheng Huang <hbisheng@gmail.com>

hhwyt committed:
Optimize flow controller minimum delay for smooth performance (#18619)

close tikv/tikv#18625

Set the flow controler minimum delay from 100ms to 0ms

The 100ms minimum delay caused severe latency jitter for a TiKV write (usually < 1ms). In this PR, the minimum delay is removed, and the write speed is now only controled by the throughput limit (`speed_limit`) in the flow cotroller.

Signed-off-by: hhwyt <hhwyt1@gmail.com>

exit-code-1 committed:
config: Add Grafana table panels for TiKV configuration metrics (#18610)

ref tikv/tikv#10047

Registered TiKV configuration metrics in Prometheus have been added to Grafana as table panels

Signed-off-by: exit-code-1 <20194550@qq.com>

hbisheng committed:
import: track inflight ingests during write stall check (#18526)

close tikv/tikv#18452

This patch addresses #18452 by serializing the write stall check using 
a mutex and tracking inflight WriteCF SSTs. This ensures all concurrent
ingests are accounted for, making the L0 check more conservative and 
preventing over-admission.

Signed-off-by: Bisheng Huang <hbisheng@gmail.com>

v01dstar committed:
Address comments

Signed-off-by: Yang Zhang <yang.zhang@pingcap.com>

v01dstar committed:
Add comment

Signed-off-by: Yang Zhang <yang.zhang@pingcap.com>

v01dstar committed:
Address comments

Signed-off-by: Yang Zhang <yang.zhang@pingcap.com>
